VirtoCommerce.CoreModule.Web/BackgroundJobs/IndexingJobs.cs

```diff
@@ -87,14 +87,15 @@ public async Task IndexAllDocumentsJob(string userName, string notificationId, I
             {
                 try
                 {
-                    if (await RunIndexJobAsync(userName, notificationId, false, o, IndexAllDocumentsAsync, cancellationToken))
-                    {
-                        // Indexation manager might re-use the jobs to scale out indexation.
-                        // Wait for all indexing jobs to complete, before telling interceptors we're ready.
-                        // This method is running as a job as well, so skip this job.
-                        // Scale out background indexation jobs are scheduled as low.
-                        await WaitForIndexationJobsToBeReadyAsync(JobPriority.Low, x => x.Method != _manualIndexAllJobMethod && x.Method != _indexChangesJobMethod);
-                    }
+                    var success = await RunIndexJobAsync(userName, notificationId, false, o, IndexAllDocumentsAsync, cancellationToken);
+
+                    // Indexation manager might re-use the jobs to scale out indexation.
+                    // Wait for all indexing jobs to complete, before telling interceptors we're ready.
+                    // This method is running as a job as well, so skip this job.
+                    // Scale out background indexation jobs are scheduled as low.
+                    await WaitForIndexationJobsToBeReadyAsync(JobPriority.Low, x => x.Method != _manualIndexAllJobMethod && x.Method != _indexChangesJobMethod);
```

Comment:
If we have result parameter success for RunIndexJobAsync method, i think we should leave the condition 
if (success)
{
 …
}
I see that in current case RunIndexJobAsync  could only succeed or fail with exception, but i do not think we should depend on that method inner logic in places where we use it.

Reply:
We should wait for all spawned jobs to complete whether success is true or not. So that the end of this task is really the end of the indexation. Is potentially important for the indexing interceptors.

Reply:
If we were unable to start the job (e.g. cannot acquire lock in RunIndexJobAsync), i think we should not wait while another indexation process will end.

Reply:
We're not waiting for another indexation process, but for spawned indexation tasks of this process.

Reply:
Sorry for late late reply. As i understand, hangfire indexation jobs could be distributed between multiple servers. If we could not acquire lock because another server performs indexation - we will wait for its indexation completion before returning `false` result, while we could return 'false' without waiting. 
Does this situation make sense?

Reply:
This waiting doesn't have anything to do with starting the indexation, but only with the end. So I don't think your solution makes sense.

Reply:
Let me describe one more situation to make clear what problem I'm talking about. Let's assume scenario with one server instance - we pressed "Build index", thus started Indexation. It would last, say, 5 minutes. After starting first index build, every second we are pressing "Build index" again. Each time we get "Indexation is already in progress" message - it is OK, but after that thread created for handling "Build index" request waits for ending of indexation. For 100 tries of pressing "Build index" button we will have 100 threads waiting for indexation completion, and we could be out of threads for handling another requests. That's why i'm asking if we could add this condition to not wait for indexation finish if it is already in progress. If it would work another way in that situation - we do not need this conditions. 
Also maybe there are some reasons when this condition brakes desired behavior I cannot see.

Reply:
I see what you mean. Thanks for your thorough explanation.
We would not use 100 threads, because the jobs would get scheduled, so parallel jobs will be limited by Hangfire. However, it would take quite a while for the user to get back response.
The reason I changed this, is so that the interceptors are only ran when all jobs of this indexation are ready.
Do I need to restructure for your use case?

Reply:
Now I understood why you removed these conditions. As it is not likely scenario, let's leave it that way.